aspnet/security/authorization/secure-data.rst

```diff
@@ -0,0 +1,369 @@
+
+.. literalinclude::  secure-data/samples/final/Data/SeedData.cs
+  :language: c#
+  :start-after: #region snippet_Contact
+  :end-before: #endregion
+  :dedent: 12
+  :emphasize-lines: 9
+
+Delete all the records in the ``Contact`` table and restart the app to seed the database. If you're using Visual Studio you might need to stop IIS Express to force the seed initializer to run.
+
+Resource based authorization
+-----------------------------
+
+- Create an *Authorization* folder for the handlers and classes we will create to implement authorization.
+- Create a *ContactIsOwnerAuthorizationHandler* class we can invoke to verify the user acting on the resource owns the resource. Create this in the *Authorization* folder.
+
+.. literalinclude:: secure-data/samples/final/Authorization/ContactIsOwnerAuthorizationHandler.cs
+  :language: c#
+
+The ``ContactIsOwnerAuthorizationHandler`` returns ``Succeed`` if the user is the contact owner. We're not checking the requirement parameter; an owner can perform any requirement on data they own.
+
+Services using Entity Framework Core must be registered for :ref:`dependency injection <fundamentals-dependency-injection>` using :dn:method:`~Microsoft.Extensions.DependencyInjection.ServiceCollectionServiceExtensions.AddScoped`. The ``ContactIsOwnerAuthorizationHandler`` uses ASP.NET Core Identity, which is built on Entity Framework Core. Register the ``ContactIsOwner`` handler with the service collection so it will be available to the ``ContactsController`` through :ref:`dependency injection <fundamentals-dependency-injection>`. Add the following code to the end of ``ConfigureServices``:
+
+.. literalinclude::  secure-data/samples/final/Startup.cs
+  :language: c#
+  :start-after: #region snippet_AddScoped
+  :end-before: #endregion
+  :dedent: 12
+
+Update the ``ContactsController`` constructor to resolve the ``IAuthorizationService`` service so we'll have access to our authorization handlers we have registered. While we're at it we'll also get the ``Identity`` ``UserManager`` service:
+
+.. literalinclude:: secure-data/samples/final/Controllers/ContactsController.cs
+  :language: c#
+  :start-after: #region snippet_ContactsController
+  :end-before: #endregion
+  :dedent: 4
+  :emphasize-lines: 4,5,9,10,13,14
+
+Add a ``ContactOperationsRequirements`` class to the *Authorization* folder to contain the requirements our app supports:
+
+.. literalinclude:: secure-data/samples/final/Authorization/ContactOperationsRequirements.cs
+  :language: c#
+
+Update the ``HTTP POST Create`` method to add the user ID to the ``Contact`` model:
+
+.. literalinclude:: secure-data/samples/final/Controllers/ContactsController.cs
+  :language: c#
+  :start-after: #region snippet_Create
+  :end-before: #endregion
+  :dedent: 8
+  :emphasize-lines: 8
+
+Update both ``Edit`` methods to use the authorization filter to verify the user owns the contact. Add ``OwnerID`` to the ``Bind`` list:
```

Comment:
handler, not filter,

Probably be worth adding something that says "As we are performing resource authorization we cannot use the normal Authorize attribute as we don't have access to the resource when attributes are evaluated. Resource based authorization must be imperative. Checks must be performed once we have access to the resource, either by loading it in our controller, or by loading it within the handler itself, passing in, for example, a unique key".